// koa-session/dist/commonjs/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Session = void 0;
const node_util_1 = require("node:util");
/**
 * Session model
 */
class Session {
    #sessCtx;
    #ctx;
    #externalKey;
    isNew = false;
    _requireSave = false;
    // session expire time, will be set from sessionData
    _expire;
    constructor(sessionContext, sessionData, externalKey) {
        this.#sessCtx = sessionContext;
        this.#ctx = sessionContext.ctx;
        this.#externalKey = externalKey;
        if (!sessionData) {
            this.isNew = true;
        }
        else {
            for (const k in sessionData) {
                // restore maxAge from store
                if (k === '_maxAge') {
                    this.#ctx.sessionOptions.maxAge = sessionData._maxAge;
                }
                else if (k === '_session') {
                    // set maxAge to 'session' if it's a session lifetime
                    this.#ctx.sessionOptions.maxAge = 'session';
                }
                else {
                    Reflect.set(this, k, sessionData[k]);
                }
            }
        }
    }
    /**
     * JSON representation of the session.
     */
    toJSON() {
        const obj = {};
        for (const key in this) {
            if (key === 'isNew')
                continue;
            // skip private stuff
            if (key[0] === '_')
                continue;
            const value = this[key];
            // skip functions
            if (typeof value === 'function')
                continue;
            obj[key] = value;
        }
        return obj;
    }
    /**
     * alias to `toJSON`
     */
    [node_util_1.inspect.custom]() {
        return this.toJSON();
    }
    /**
     * Return how many values there are in the session object.
     * Used to see if it's "populated".
     */
    get length() {
        return Object.keys(this.toJSON()).length;
    }
    /**
     * populated flag, which is just a boolean alias of .length.
     */
    get populated() {
        return !!this.length;
    }
    /**
     * get session maxAge
     */
    get maxAge() {
        return this.#ctx.sessionOptions.maxAge;
    }
    /**
     * set session maxAge
     */
    set maxAge(val) {
        this.#ctx.sessionOptions.maxAge = val;
        // maxAge changed, must save to cookie and store
        this._requireSave = true;
    }
    /**
     * get session external key
     * only exist if opts.store present
     */
    get externalKey() {
        return this.#externalKey;
    }
    /**
     * save this session no matter whether it is populated
     *
     * @param {Function} [callback] the optional function to call after saving the session
     */
    save(callback) {
        return this.commit({ save: true }, callback);
    }
    /**
     * regenerate this session
     *
     * @param  {Function} [callback] the optional function to call after regenerating the session
     */
    regenerate(callback) {
        return this.commit({ regenerate: true }, callback);
    }
    /**
     * commit this session's headers if autoCommit is set to false
     */
    manuallyCommit() {
        return this.commit();
    }
    commit(options, callback) {
        if (typeof options === 'function') {
            callback = options;
            options = {};
        }
        const promise = this.#sessCtx.commit(options);
        if (callback) {
            promise.then(() => callback(), callback);
        }
        else {
            return promise;
        }
    }
}
exports.Session = Session;
