// feathers-authentication-management/dist/methods/verify-signup-set-password.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifySignupSetPasswordWithShortToken = exports.verifySignupSetPasswordWithLongToken = void 0;
const errors_1 = require("@feathersjs/errors");
const debug_1 = __importDefault(require("debug"));
const helpers_1 = require("../helpers");
const debug = (0, debug_1.default)('authLocalMgnt:verifySignupSetPassword');
function verifySignupSetPasswordWithLongToken(options, verifyToken, password, notifierOptions = {}, params) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, helpers_1.ensureValuesAreStrings)(verifyToken, password);
        const result = yield verifySignupSetPassword(options, { verifyToken }, { verifyToken }, password, notifierOptions, params);
        return result;
    });
}
exports.verifySignupSetPasswordWithLongToken = verifySignupSetPasswordWithLongToken;
function verifySignupSetPasswordWithShortToken(options, verifyShortToken, identifyUser, password, notifierOptions = {}, params) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, helpers_1.ensureValuesAreStrings)(verifyShortToken, password);
        (0, helpers_1.ensureObjPropsValid)(identifyUser, options.identifyUserProps);
        const result = yield verifySignupSetPassword(options, identifyUser, {
            verifyShortToken
        }, password, notifierOptions, params);
        return result;
    });
}
exports.verifySignupSetPasswordWithShortToken = verifySignupSetPasswordWithShortToken;
function verifySignupSetPassword(options, identifyUser, tokens, password, notifierOptions = {}, params) {
    return __awaiter(this, void 0, void 0, function* () {
        debug('verifySignupSetPassword', identifyUser, tokens, password);
        if (params && "query" in params) {
            params = Object.assign({}, params);
            delete params.query;
        }
        const { app, passwordField, skipPasswordHash, sanitizeUserForClient, service, notifier } = options;
        const usersService = app.service(service);
        const usersServiceId = usersService.id;
        const users = yield usersService.find(Object.assign(Object.assign({}, params), { query: Object.assign(Object.assign({}, identifyUser), { $limit: 2 }), paginate: false }));
        const user = (0, helpers_1.getUserData)(users, [
            'isNotVerifiedOrHasVerifyChanges',
            'verifyNotExpired'
        ]);
        if (!Object.keys(tokens).every((key) => tokens[key] === user[key])) {
            yield eraseVerifyProps(user, user.isVerified, params);
            throw new errors_1.BadRequest('Invalid token. Get for a new one. (authLocalMgnt)', { errors: { $className: 'badParam' } });
        }
        const userErasedVerify = yield eraseVerifyPropsSetPassword(user, (0, helpers_1.isDateAfterNow)(user.verifyExpires), user.verifyChanges || {}, password, skipPasswordHash, params);
        const userResult = yield (0, helpers_1.notify)(notifier, 'verifySignupSetPassword', userErasedVerify, notifierOptions);
        return sanitizeUserForClient(userResult);
        function eraseVerifyProps(user, isVerified, params) {
            return __awaiter(this, void 0, void 0, function* () {
                const patchData = Object.assign({}, {
                    isVerified,
                    verifyToken: null,
                    verifyShortToken: null,
                    verifyExpires: null,
                    verifyChanges: {},
                });
                const result = yield usersService.patch(user[usersServiceId], patchData, Object.assign({}, params));
                return result;
            });
        }
        function eraseVerifyPropsSetPassword(user, isVerified, verifyChanges, password, skipPasswordHash, params) {
            return __awaiter(this, void 0, void 0, function* () {
                const patchData = Object.assign({}, verifyChanges || {}, {
                    isVerified,
                    verifyToken: null,
                    verifyShortToken: null,
                    verifyExpires: null,
                    verifyChanges: {},
                    [passwordField]: skipPasswordHash ? password : yield (0, helpers_1.hashPassword)(app, password, passwordField)
                });
                const result = yield usersService.patch(user[usersServiceId], patchData, Object.assign({}, params));
                return result;
            });
        }
    });
}
