// feathers-authentication-management/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.client = exports.VerifySignupShortService = exports.VerifySignupSetPasswordShortService = exports.VerifySignupSetPasswordLongService = exports.VerifySignupLongService = exports.SendResetPwdService = exports.ResetPwdShortService = exports.ResetPwdLongService = exports.ResendVerifySignupService = exports.PasswordChangeService = exports.IdentityChangeService = exports.CheckUniqueService = exports.AuthenticationManagementService = exports.removeVerification = exports.isVerified = exports.addVerification = exports.hooks = void 0;
const setupAuthManagement_1 = __importDefault(require("./setupAuthManagement"));
const hooks_1 = require("./hooks");
Object.defineProperty(exports, "addVerification", { enumerable: true, get: function () { return hooks_1.addVerification; } });
Object.defineProperty(exports, "isVerified", { enumerable: true, get: function () { return hooks_1.isVerified; } });
Object.defineProperty(exports, "removeVerification", { enumerable: true, get: function () { return hooks_1.removeVerification; } });
exports.default = setupAuthManagement_1.default;
exports.hooks = {
    addVerification: hooks_1.addVerification,
    isVerified: hooks_1.isVerified,
    removeVerification: hooks_1.removeVerification
};
var AuthenticationManagementService_1 = require("./services/AuthenticationManagementService");
Object.defineProperty(exports, "AuthenticationManagementService", { enumerable: true, get: function () { return AuthenticationManagementService_1.AuthenticationManagementService; } });
var CheckUniqueService_1 = require("./services/CheckUniqueService");
Object.defineProperty(exports, "CheckUniqueService", { enumerable: true, get: function () { return CheckUniqueService_1.CheckUniqueService; } });
var IdentityChangeService_1 = require("./services/IdentityChangeService");
Object.defineProperty(exports, "IdentityChangeService", { enumerable: true, get: function () { return IdentityChangeService_1.IdentityChangeService; } });
var PasswordChangeService_1 = require("./services/PasswordChangeService");
Object.defineProperty(exports, "PasswordChangeService", { enumerable: true, get: function () { return PasswordChangeService_1.PasswordChangeService; } });
var ResendVerifySignupService_1 = require("./services/ResendVerifySignupService");
Object.defineProperty(exports, "ResendVerifySignupService", { enumerable: true, get: function () { return ResendVerifySignupService_1.ResendVerifySignupService; } });
var ResetPwdLongService_1 = require("./services/ResetPwdLongService");
Object.defineProperty(exports, "ResetPwdLongService", { enumerable: true, get: function () { return ResetPwdLongService_1.ResetPwdLongService; } });
var ResetPwdShortService_1 = require("./services/ResetPwdShortService");
Object.defineProperty(exports, "ResetPwdShortService", { enumerable: true, get: function () { return ResetPwdShortService_1.ResetPwdShortService; } });
var SendResetPwdService_1 = require("./services/SendResetPwdService");
Object.defineProperty(exports, "SendResetPwdService", { enumerable: true, get: function () { return SendResetPwdService_1.SendResetPwdService; } });
var VerifySignupLongService_1 = require("./services/VerifySignupLongService");
Object.defineProperty(exports, "VerifySignupLongService", { enumerable: true, get: function () { return VerifySignupLongService_1.VerifySignupLongService; } });
var VerifySignupSetPasswordLongService_1 = require("./services/VerifySignupSetPasswordLongService");
Object.defineProperty(exports, "VerifySignupSetPasswordLongService", { enumerable: true, get: function () { return VerifySignupSetPasswordLongService_1.VerifySignupSetPasswordLongService; } });
var VerifySignupSetPasswordShortService_1 = require("./services/VerifySignupSetPasswordShortService");
Object.defineProperty(exports, "VerifySignupSetPasswordShortService", { enumerable: true, get: function () { return VerifySignupSetPasswordShortService_1.VerifySignupSetPasswordShortService; } });
var VerifySignupShortService_1 = require("./services/VerifySignupShortService");
Object.defineProperty(exports, "VerifySignupShortService", { enumerable: true, get: function () { return VerifySignupShortService_1.VerifySignupShortService; } });
__exportStar(require("./types"), exports);
var client_1 = require("./client");
Object.defineProperty(exports, "client", { enumerable: true, get: function () { return __importDefault(client_1).default; } });
// commonjs
if (typeof module !== 'undefined') {
    module.exports = Object.assign(setupAuthManagement_1.default, module.exports);
}
