// feathers-authentication-management/dist/options.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeDefaultOptions = exports.optionsDefault = exports.defaultPath = void 0;
const cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
const pick_1 = __importDefault(require("lodash/pick"));
const sanitize_user_for_client_1 = require("./helpers/sanitize-user-for-client");
exports.defaultPath = 'authManagement';
exports.optionsDefault = {
    service: '/users',
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
    notifier: (type, user, notifierOptions) => __awaiter(void 0, void 0, void 0, function* () { }),
    longTokenLen: 15,
    shortTokenLen: 6,
    shortTokenDigits: true,
    resetDelay: 1000 * 60 * 60 * 2,
    delay: 1000 * 60 * 60 * 24 * 5,
    resetAttempts: 0,
    reuseResetToken: false,
    identifyUserProps: ['email'],
    sanitizeUserForClient: sanitize_user_for_client_1.sanitizeUserForClient,
    skipIsVerifiedCheck: false,
    passwordField: 'password',
    skipPasswordHash: false,
    passParams: undefined
};
function makeDefaultOptions(keys) {
    const options = (0, cloneDeep_1.default)(exports.optionsDefault);
    if (!keys) {
        return options;
    }
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    return (0, pick_1.default)(options, keys);
}
exports.makeDefaultOptions = makeDefaultOptions;
